File: app/graphql/mutations/sign_up.rb
module Mutations
  class SignUp < BaseMutation
    description 'Sign up as a new user'
    payload_type Types::UserType
    null true

    argument :name, String, required: true
    argument :email, String, required: true
    argument :password, String, required: true
    argument :password_confirmation, String, required: true

    def resolve(name:, email:, password:, password_confirmation:)
      result = CreateUser.call(
        name: name,
        email: email,
        password: password,
        password_confirmation: password_confirmation
      )

      if result.success?
        context[:environment].current_user = result.user
      end
    end
  end
end

File: app/javascript/queries/mutations/sign_up.js
import gql from 'graphql-tag'
import { CURRENT_USER_FIELDS } from '../current_user'

export default gql`
  mutation signUpMutation(
    $name: String!,
    $email: String!,
    $password: String!,
    $passwordConfirmation: String!
  ) {
    signUp(
      name: $name,
      email: $email,
      password: $password,
      passwordConfirmation: $passwordConfirmation
    ) {
      ...CurrentUserFields
    }
  }

  ${CURRENT_USER_FIELDS}
`

